File: Simple_Terr_starter.m
% Read monthly surface forcing data for simple terrestrial ecosystem model
% Matlab script
% Mick Follows, Oliver Jahn, Mukund Gupta, Deepa Rao (2019)

% clear all data and close all figures
clear all; 
close all;

% select month using strings
% mon = '01' for january etc. 
% mon = 'mean' for annual mean
%
% create an array of strings representing months
months = {'01','02','03','04','05','06','07','08','09','10','11','12'} ; 

% set mon to January ...
mon = char(months(1)) ;
% uncomment next line to use annual mean
%mon = 'mean' ;

% read in the lat and lon coordinates of the gridded data
lat = importdata('latitude.txt');
lon = importdata('longitude.txt');
nlat = length(lat);
nlon = length(lon);

% initialize array (lat, lon, month)
T = zeros(length(lat),length(lon),12);
sw = zeros(length(lat),length(lon),12);
precip = zeros(length(lat),length(lon),12);

% Reading data for monthly means
for i=1:length(months)
    mon = months{i};
    % surface air temperature (C)
    T(:,:,i) = importdata(['SurfTair_' mon '.txt']);
    % precipitation (cm/month)
    precip(:,:,i) = importdata(['Precip_' mon '.txt']);
    % incident short wave radiation - visible light (W/m2)
    sw(:,:,i) = importdata(['SurfSW_' mon '.txt']);
end

% convert precip from cm/month to m/yr 
precip = precip * 12.0 / 100.0; % Converting to [m/yr]

% mask missing values
T(T>1e36) = nan; % set missing values to NaNs 
precip(precip>1e36) = nan; % set missing values to NaNs 
sw(sw>1e36) = nan; % set missing values to NaNs 

% Make a mask setting ocean points to NaN based on the T file
% Divide T by itself - will give 1s and NaNs
landmask = T ./ T ;

% mask out ocean points in the shortwave file
sw = sw .* landmask; 

%plot surface temperature (C)
figure(1);
pcolor(lon, lat, T(:,:,1))
shading flat
colorbar
xlabel('Longitude', 'fontsize', 14)
ylabel('Latitude', 'fontsize', 14)
title(['Temperature (^oC)'], 'fontsize', 16)

File: Simple_Terr_starter.py
# Read monthly surface forcing data for simple terrestrial ecosystem model
# Python script
# Mick Follows, Deepa Rao (2019)
# Class exercise, Feb 2019 

# For use in Jupyter Notebook include %matplotlib inline

import os #functions for working with file system
import sys
import glob
from pylab import *

import scipy as sp
import numpy as np
import numpy.ma as ma #masked arrays
import pandas as pd
import matplotlib.pyplot as plt
# %matplotlib inline

# print the current working directory
os.getcwd()
# change working directory to where input files are
os.chdir('/Users/Deepa/Documents/12.349:12.849/Terrestrial Biosphere/monthly-input-data')
# Backslash on Windows and Forward Slash on OS X and Linux

# read in the lat and lon coordinates of gridded data
lat = pd.read_table('latitudegrid.txt', header=None, skiprows=1)
lon = pd.read_table('longitudegrid.txt', header=None, skiprows=1)

# create meshgrid for plotting of lat and lon
lat, lon = np.meshgrid(lat.T,lon.T) #create meshgrid for plotting 

# import monthly avg climatology data as a numpy array

# use glob.glob to get an unordered list of filenames ('prefix_*.txt')
#print (glob.glob('*.txt')) #print all files in directory
T_files = glob.glob('SurfTair_*.txt')
precip_files = glob.glob('Precip_*.txt') 
sw_files = glob.glob('SurfSW_*.txt') 

# sort unordered list by filename (01, 02, 03 ...)
T_files.sort()
precip_files.sort()
sw_files.sort()

#  check files and array shape
#print(precip_files)
#print(lat.T.shape, lon.T.shape, T_Jan.shape)

# import monthly avg climatology data as a numpy array and mask missing values (1e36)
T = ma.masked_greater([np.loadtxt(f) for f in T_files], 1e36) #surface air temp (C)
precip = ma.masked_greater([np.loadtxt(f) for f in precip_files], 1e36) #precipitation (cm/month)
precip = precip * 12.0 / 100.0 #convert from (cm/month) to (m/yr)
sw = ma.masked_greater([np.loadtxt(f) for f in sw_files], 1e36) #incident short wave radiation - visible light (W m-2)

# test plot -- January temperature
plt.contourf(lon,lat,T[0].T) #note: transpose matrix to set matrix dimensions for plotting
plt.contourf(lon,lat, T[1].T, cmap = cm.coolwarm)
plt.title('Feb mean Surface Air Tempearture (C)')

# plot monthly T 
fig, ax = plt.subplots()
CS = ax.contourf(lon, lat, T[0].T, cmap=cm.coolwarm)
ax.set(xlim=[-180, 180], ylim=[-90,90],
       xlabel='Longitude', ylabel='Latitude',
       title='Monthly Mean Suface Air Temperature (C)')

# create landmask based on surface air T
landmask = np.ma.masked_where(T[0]>1e36, T[0])

# apply landmask on monthly sw data -- to only plot terrestrial data
sw_mask_0 = np.ma.masked_where(np.ma.getmask(landmask), sw[0]) #applies landmask on monthly[n] sw data

# plot monthly SW with landmask
fig, ax = plt.subplots()
J = ax.contourf(lon, lat, sw_mask_0.T, cmap=cm.coolwarm)
ax.set(xlim=[-180, 180], ylim=[-90,90],
       xlabel='Longitude', ylabel='Latitude',
       title='Jan SW (W/m2)')
    